exporter/exporterhelper/queue.go

```diff
@@ -0,0 +1,104 @@
+func NewDefaultQueueConfig() QueueConfig {
+	return QueueConfig{
+		Enabled:        true,
+		NumConsumers:   10,
+		QueueSizeItems: 100_000,
+	}
+}
+
+// Validate checks if the QueueSettings configuration is valid
+func (qCfg *QueueConfig) Validate() error {
+	if !qCfg.Enabled {
+		return nil
+	}
+	if qCfg.NumConsumers <= 0 {
+		return errors.New("number of consumers must be positive")
+	}
+	if qCfg.QueueSizeItems > 0 && qCfg.QueueSize > 0 {
+		return errors.New("only one of 'queue_size' and 'queue_size_items' can be specified")
+	}
+	if qCfg.QueueSizeItems <= 0 && qCfg.QueueSize <= 0 {
+		return errors.New("queue size must be positive")
+	}
+	return nil
+}
+
+// Queue defines a producer-consumer exchange which can be backed by e.g. the memory-based ring buffer queue
+// (boundedMemoryQueue) or via a disk-based queue (persistentQueue)
+// This API is at the early stage of development and may change without backward compatibility
+// until https://github.com/open-telemetry/opentelemetry-collector/issues/8122 is resolved.
+type Queue = internal.Queue[Request]
+
+// QueueCreateSettings defines settings for creating a queue.
+type QueueCreateSettings = internal.QueueCreateSettings[Request]
+
+type QueueFactory func(QueueCreateSettings) Queue
+
+// NewMemoryQueueFactory returns a factory to create a new memory queue.
+// This API is at the early stage of development and may change without backward compatibility
+// until https://github.com/open-telemetry/opentelemetry-collector/issues/8122 is resolved.
+func NewMemoryQueueFactory() QueueFactory {
+	return func(set QueueCreateSettings) Queue {
+		return internal.NewBoundedMemoryQueue[Request](set)
+	}
+}
+
+// PersistentQueueConfig defines configuration for queueing requests in a persistent storage.
+// The struct is provided to be added in the exporter configuration as one struct under the "sending_queue" key.
+// The exporter helper Go interface requires the fields to be provided separately to WithRequestQueue and
+// NewPersistentQueueFactory.
+// This API is at the early stage of development and may change without backward compatibility
+// until https://github.com/open-telemetry/opentelemetry-collector/issues/8122 is resolved.
+type PersistentQueueConfig struct {
+	QueueConfig `mapstructure:",squash"`
+	// StorageID if not empty, enables the persistent storage and uses the component specified
+	// as a storage extension for the persistent queue
+	StorageID *component.ID `mapstructure:"storage"`
+}
+
+// NewPersistentQueueFactory returns a factory to create a new persistent queue.
+// If cfg.StorageID is nil then it falls back to memory queue.
+// This API is at the early stage of development and may change without backward compatibility
+// until https://github.com/open-telemetry/opentelemetry-collector/issues/8122 is resolved.
+func NewPersistentQueueFactory(storageID *component.ID, marshaler RequestMarshaler, unmarshaler RequestUnmarshaler) QueueFactory {
```

Comment:
Provide a "PersistentQueueSettings" to allow extending. Include the Marshaler/Unmarshaler for the moment.

Reply:
Makes sense. Applied